  
<h1 align="center">Capstone Project – The Battle of Neighborhoods</h1>
<h2 align="center">The Place to Be</h2>

Akhil Sachar

# Introduction

The place to be is the project for the prople who wish to move to a neighborhood they have no knowledge about. They could be students who wish to persue higher education, immigrants from different states of Canada or oll over the world. I hope this will work wonders for them.

## Data Description:

We using the data from wikipedia to find postal codes of different places.
Data Link: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M 

Will use dataset which we scrapped from wikipedia. Dataset consisting of latitude and longitude, zip codes which we got from the course

In [93]:
import pandas as pd
import requests
import numpy as np
import geocoder
import folium
import requests 
import matplotlib.cm as cm
import matplotlib.colors as colors
import json
import xml
import matplotlib.pyplot as plt
from bs4 import BeautifulSoup
from geopy.geocoders import Nominatim 
from sklearn.cluster import KMeans


In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Response 200 means connection successful

In [29]:

soup= BeautifulSoup(wiki_url.text)

In [51]:

jack=pd.DataFrame(columns=["PostalCode", "Borough","Neighborhood"])

Converting HTML text into dataframe

In [52]:

table=soup.find('table')
n=0
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        jack.loc[n,'PostalCode']=cell['PostalCode']
        jack.loc[n,'Borough']=cell['Borough']
        jack.loc[n,'Neighborhood']=cell['Neighborhood']
        
        n=n+1
        

In [32]:

jack

PostalCode                                            Borough  \
0          M3A                                         North York   
1          M4A                                         North York   
2          M5A                                   Downtown Toronto   
3          M6A                                         North York   
4          M7A                                       Queen's Park   
..         ...                                                ...   
98         M8X                                          Etobicoke   
99         M4Y                                   Downtown Toronto   
100        M7Y  East TorontoBusiness reply mail Processing Cen...   
101        M8Y                                          Etobicoke   
102        M8Z                                          Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [53]:
jack.shape

(103, 3)

There are 103 entries in the dataframe

Now we will upload LAttitude and Longitude data to the project

In [54]:
data = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

Check shape of both dataframes

In [55]:
print("The shape of our wiki data is: ", jack.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Chect their Data types

In [56]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [57]:
jack.dtypes

PostalCode      object
Borough         object
Neighborhood    object
dtype: object

Now we shall combine them

In [76]:
combined_data = jack.join(data.set_index('Postal Code'), on='PostalCode', how='inner')
combined_data

PostalCode                                            Borough  \
0          M3A                                         North York   
1          M4A                                         North York   
2          M5A                                   Downtown Toronto   
3          M6A                                         North York   
4          M7A                                       Queen's Park   
..         ...                                                ...   
98         M8X                                          Etobicoke   
99         M4Y                                   Downtown Toronto   
100        M7Y  East TorontoBusiness reply mail Processing Cen...   
101        M8Y                                          Etobicoke   
102        M8Z                                          Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                        Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100                                     Enclave of M4L  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [67]:
combined_data.shape

(103, 5)

In [78]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [38]:
combined_data['Borough'].value_counts()

North York                                                      24
Downtown Toronto                                                17
Scarborough                                                     17
Etobicoke                                                       11
Central Toronto                                                  9
West Toronto                                                     6
York                                                             5
East Toronto                                                     4
East York                                                        4
EtobicokeNorthwest                                               1
Queen's Park                                                     1
East TorontoBusiness reply mail Processing Centre969 Eastern     1
MississaugaCanada Post Gateway Processing Centre                 1
East YorkEast Toronto                                            1
Downtown TorontoStn A PO Boxes25 The Esplanade                

### Map of toronto

In [117]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighborhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

### Four Square API

In [73]:
CLIENT_ID = '15BOMRSBES0DKU0LQGMP5YXQJ4VMP1NW3IKX31HASULDF1W2' 
CLIENT_SECRET = 'HQAB2PM2C2GTWUHTDO1HY4KMJS2ORY4GPRWEDLAR20AWMWSD'
VERSION = '20180601' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 15BOMRSBES0DKU0LQGMP5YXQJ4VMP1NW3IKX31HASULDF1W2
CLIENT_SECRET:HQAB2PM2C2GTWUHTDO1HY4KMJS2ORY4GPRWEDLAR20AWMWSD


In [81]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Getting all the venues in Toronto

In [82]:
venues_in_toronto = getNearbyVenues(combined_data['Neighborhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [83]:
venues_in_toronto.shape

(1323, 5)

In [84]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

In order for us to find out best neighbourhoods to be in we will need to group these by Neighbourhood.

In [85]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1311  Mimico NW, The Queensway West, South of Bloor,...   
1312  Mimico NW, The Queensway West, South of Bloor,...   
1313  Mimico NW, The Queensway West, South of Bloor,...   
1314  Mimico NW, The Queensway West, South of Bloor,...   
1315  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572               Portugril   
4                  43.725882               -79.315572             Tim Hortons   
...                      ...                      ...                     ...   
1311               43.628841               -79.520999        South St. Burger   
1312               43.628841               -79.520999              Wingporium   
1313               43.628841               -79.520999               Dollarama   
1314               43.628841               -79.520999          Healthy Planet   
1315               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3     Portuguese Restaurant  
4               Coffee Shop  
...                     ...  
1311           Burger Joint  
1312            Wings Joint  
1313         Discount Store  
1314        Supplement Shop  
1315                 Bakery  

[429 rows x 5 columns]

Checking the venues by higher chance of occurance 

In [86]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                                   Wexford, Maryvale   
Adult Boutique                                   Church and Wellesley   
Airport                                                Downsview East   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.750071               -79.295849   
Adult Boutique                   43.665860               -79.383160   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store                       Puffin Gear  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[239 rows x 4 columns]

### One Hot Encoding

In [87]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1318                  0               0        0                   0   
1319                  0               0        0                   0   
1320                  0               0        0                   0   
1321                  0               0        0                   0   
1322                  0               0        0                   0   

      Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0                0               0                0                 0   
1                0               0                0                 0   
2                0               0                0                 0   
3                0               0                0                 0   
4                0               0                0                 0   
...            ...             ...              ...               ...   
1318             0               0                0                 0   
1319             0               0                0                 0   
1320             0               0                0                 0   
1321             0               0                0                 0   
1322             0               0                0                 0   

      American Restaurant  Antique Shop  ...  Turkish Restaurant  \
0                       0             0  ...                   0   
1                       0             0  ...                   0   
2                       0             0  ...                   0   
3                       0             0  ...                   0   
4                       0             0  ...                   0   
...                   ...           ...  ...                 ...   
1318                    0             0  ...                   0   
1319                    0             0  ...                   0   
1320                    0             0  ...                   0   
1321                    0             0  ...                   0   
1322                    0             0  ...                   0   

      Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                                 0                 0            0   
1                                 0                 0            0   
2                                 0                 0            0   
3                                 0                 0            0   
4                                 0                 0            0   
...                             ...               ...          ...   
1318                              0                 0            0   
1319                              0                 0            0   
1320                              0                 0            0   
1321                              0                 0            0   
1322                              0                 0            0   

      Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                         0                0         0            0   
1                         0                0         0            0   
2                         0                0         0            0   
3                         0                0         0            0   
4                         0                0         0            0   
...                     ...              ...       ...          ...   
1318                      0                0         0            0   
1319                      0           

Adding Neighbourhood to the dataframe

In [88]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2  Victoria Village                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Turkish Restaurant  \
0                 0                    0  ...                   0   
1                 0                    0  ...                   0   
2                 0                    0  ...                   0   
3                 0                    0  ...                   0   
4                 0                    0  ...                   0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 240 columns]

In [89]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0             0.0      0.0                 0.0           0.0             0.0   
1             0.0      0.0                 0.0           0.0             0.0   
2             0.0      0.0                 0.0           0.0             0.0   
3             0.0      0.0                 0.0           0.0             0.0   
4             0.0      0.0                 0.0           0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  ...  \
0              0.0               0.0             0.000000  ...   
1              0.0               0.0             0.000000  ...   
2              0.0               0.0             0.000000  ...   
3              0.0               0.0             0.000000  ...   
4              0.0               0.0             0.038462  ...   

   Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                 0.0                            0.0               0.0   
1                 0.0                            0.0               0.0   
2                 0.0                            0.0               0.0   
3                 0.0                            0.0               0.0   
4                 0.0                            0.0               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0          0.0                    0.0              0.0       0.0          0.0   
1          0.0                    0.0              0.0       0.0          0.0   
2          0.0                    0.0              0.0       0.0          0.0   
3          0.0                    0.0              0.0       0.0          0.0   
4          0.0                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 240 columns]

Making a function to get higest venue category

In [90]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

The number of categories is high so we will take 10

In [91]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt        Breakfast Spot   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village   Japanese Restaurant   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge        Clothing Store          Skating Rink   
1                  Pool          Skating Rink        Sandwich Place   
2           Coffee Shop           Pizza Place                 Diner   
3                  Café                  Bank    Chinese Restaurant   
4            Restaurant           Coffee Shop    Italian Restaurant   

       5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0  Latin American Restaurant     Accessories Store                Music Venue   
1                Coffee Shop                   Pub                        Gym   
2  Middle Eastern Restaurant         Shopping Mall          Mobile Phone Shop   
3              Metro Station    Mexican Restaurant  Middle Eastern Restaurant   
4                Pizza Place      Toy / Game Store       Fast Food Restaurant   

  8th Most Common Venue      9th Most Common Venue      10th Most Common Venue  
0    Mexican Restaurant  Middle Eastern Restaurant          Miscellaneous Shop  
1              Pharmacy              Metro Station                 Men's Store  
2        Sandwich Place                Supermarket            Sushi Restaurant  
3    Miscellaneous Shop          Mobile Phone Shop  Modern European Restaurant  
4        Cosmetics Shop                    Butcher                        Café

Making the model 

In [94]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [95]:
kmeans.labels_[0:100]

array([3, 3, 3, 3, 3, 3, 3, 3, 3, 1, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
       3, 3, 3, 1, 3, 1, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 0, 3,
       3, 3, 3, 3, 3, 1, 3, 1, 3, 3, 0, 1, 3, 4, 3, 1, 3, 3, 3, 3, 3, 3,
       1, 3, 3, 1, 3, 2, 3, 3, 4, 3, 3, 3, 3, 3, 3, 3, 3, 1, 3, 3, 3, 3,
       3, 3, 3, 3, 1, 3, 3, 3, 1, 3, 3, 1])

In [96]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [97]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighborhood')

toronto_merged.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.329656             1.0     Food & Drink Shop                  Park   
1 -79.315572             3.0          Intersection           Pizza Place   
2 -79.360636             3.0           Coffee Shop                Bakery   
3 -79.464763             3.0        Clothing Store     Accessories Store   
4 -79.389494             3.0           Coffee Shop      Sushi Restaurant   

   3rd Most Common Venue 4th Most Common Venue     5th Most Common Venue  \
0      Accessories Store                Museum  Mediterranean Restaurant   
1  Portuguese Restaurant           Coffee Shop              Hockey Arena   
2                   Park                  Café            Breakfast Spot   
3  Vietnamese Restaurant           Coffee Shop        Miscellaneous Shop   
4            Yoga Studio                  Bank                  Beer Bar   

        6th Most Common Venue   7th Most Common Venue  \
0                 Men's Store           Metro Station   
1  Modern European Restaurant                  Museum   
2                 Yoga Studio    Gym / Fitness Center   
3                    Boutique  Furniture / Home Store   
4               Smoothie Shop           Burrito Place   

       8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0         Mexican Restaurant  Middle Eastern Restaurant     Miscellaneous Shop  
1              Movie Theater                      Motel    Monument / Landmark  
2          French Restaurant             Farmers Market    Distribution Center  
3  Middle Eastern Restaurant         Mexican Restaurant          Metro Station  
4                       Café             Sandwich Place             Restaurant

In [98]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on map

In [99]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [100]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighborhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

### Cluster 1

In [101]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
6   Scarborough             0.0  Fast Food Restaurant   
27   North York             0.0  Fast Food Restaurant   

       2nd Most Common Venue 3rd Most Common Venue     4th Most Common Venue  \
6             Massage Studio        Medical Center  Mediterranean Restaurant   
27  Mediterranean Restaurant                  Pool               Golf Course   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6            Men's Store         Metro Station    Mexican Restaurant   
27               Dog Run                Museum           Men's Store   

        8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
6   Middle Eastern Restaurant    Miscellaneous Shop          Mobile Phone Shop  
27              Metro Station    Mexican Restaurant  Middle Eastern Restaurant

## Cluster 2

In [102]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0              North York             1.0     Food & Drink Shop   
21                   York             1.0                  Park   
35  East YorkEast Toronto             1.0                  Park   
40             North York             1.0               Airport   
46             North York             1.0         Grocery Store   
49             North York             1.0        Massage Studio   
52             North York             1.0                  Park   
61        Central Toronto             1.0    Photography Studio   
64                   York             1.0     Convenience Store   
66             North York             1.0     Convenience Store   
77              Etobicoke             1.0        Sandwich Place   
85            Scarborough             1.0          Intersection   
91       Downtown Toronto             1.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue  \
0                         Park     Accessories Store   
21               Women's Store                   Bar   
35                 Coffee Shop     Convenience Store   
40                        Park              Bus Stop   
46                        Park         Shopping Mall   
49  Construction & Landscaping                  Park   
52           Accessories Store           Music Venue   
61                        Park           Swim School   
64                        Park     Accessories Store   
66                        Park     Accessories Store   
77                        Park     Mobile Phone Shop   
85                        Park            Playground   
91                  Playground                 Trail   

       4th Most Common Venue     5th Most Common Venue  \
0                     Museum  Mediterranean Restaurant   
21         Accessories Store                    Museum   
35         Accessories Store               Music Venue   
40         Accessories Store               Music Venue   
46                      Bank         Accessories Store   
49                    Bakery               Music Venue   
52  Mediterranean Restaurant               Men's Store   
61                  Bus Line         Accessories Store   
64               Music Venue  Mediterranean Restaurant   
66               Music Venue  Mediterranean Restaurant   
77                  Bus Line         Accessories Store   
85        Photography Studio                    Museum   
91                    Museum            Medical Center   

       6th Most Common Venue     7th Most Common Venue  \
0                Men's Store             Metro Station   
21  Mediterranean Restaurant               Men's Store   
35               Men's Store             Metro Station   
40               Men's Store             Metro Station   
46                    Museum               Men's Store   
49               Men's Store             Metro Station   
52             Metro Station        Mexican Restaurant   
61             Movie Theater  Mediterranean Restaurant   
64               Men's Store             Metro Station   
66               Men's Store             Metro Station   
77               Men's Store             Metro Station   
85  Mediterranean Restaurant               Men's Store   
91  Mediterranean Restaurant               Men's Store   

        8th Most Common Venue      9th Most Common Venue  \
0          Mexican Restaurant  Middle Eastern Restaurant   
21              Metro Station         Mexican Restaurant   
35         Mexican Restaurant  Middle Eastern Restaurant   
40         Mexican Restaurant  Middle Eastern Restaurant   
46              Metro Station         Mexican Restaurant   
49         Mexican Restaurant  Middle Eastern Restaurant   
52  Middle Eastern Restaurant         Miscellaneous Shop   
61                Men's Store              Metro Station   
64         Mexican Restaurant  Middle Eastern Restaurant   
66         Mexican Restaurant  Middle Eastern Restaurant   
77      

## Cluster 3

In [103]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
12  Scarborough             2.0                   Bar     Accessories Store   

   3rd Most Common Venue     4th Most Common Venue 5th Most Common Venue  \
12        Massage Studio  Mediterranean Restaurant           Men's Store   

   6th Most Common Venue 7th Most Common Venue      8th Most Common Venue  \
12         Metro Station    Mexican Restaurant  Middle Eastern Restaurant   

   9th Most Common Venue 10th Most Common Venue  
12    Miscellaneous Shop      Mobile Phone Shop

## Cluster 4

In [104]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels  \
1                                           North York             3.0   
2                                     Downtown Toronto             3.0   
3                                           North York             3.0   
4                                         Queen's Park             3.0   
7                                           North York             3.0   
..                                                 ...             ...   
98                                           Etobicoke             3.0   
99                                    Downtown Toronto             3.0   
100  East TorontoBusiness reply mail Processing Cen...             3.0   
101                                          Etobicoke             3.0   
102                                          Etobicoke             3.0   

    1st Most Common Venue       2nd Most Common Venue  3rd Most Common Venue  \
1            Intersection                 Pizza Place  Portuguese Restaurant   
2             Coffee Shop                      Bakery                   Park   
3          Clothing Store           Accessories Store  Vietnamese Restaurant   
4             Coffee Shop            Sushi Restaurant            Yoga Studio   
7     Japanese Restaurant                Dessert Shop                    Gym   
..                    ...                         ...                    ...   
98                  River           Accessories Store            Music Venue   
99       Sushi Restaurant              Breakfast Spot     Salon / Barbershop   
100            Comic Shop            Recording Studio         Farmers Market   
101                  Pool  Construction & Landscaping         Baseball Field   
102                Bakery             Supplement Shop         Hardware Store   

        4th Most Common Venue 5th Most Common Venue  \
1                 Coffee Shop          Hockey Arena   
2                        Café        Breakfast Spot   
3                 Coffee Shop    Miscellaneous Shop   
4                        Bank              Beer Bar   
7        Caribbean Restaurant                  Café   
..                        ...                   ...   
98   Mediterranean Restaurant           Men's Store   
99           Ramen Restaurant           Coffee Shop   
100      Fast Food Restaurant           Pizza Place   
101         Accessories Store           Music Venue   
102               Wings Joint     Convenience Store   

          6th Most Common Venue   7th Most Common Venue  \
1    Modern European Restaurant                  Museum   
2                   Yoga Studio    Gym / Fitness Center   
3                      Boutique  Furniture / Home Store   
4                 Smoothie Shop           Burrito Place   
7                 Movie Theater            Neighborhood   
..                          ...                     ...   
98                Metro Station      Mexican Restaurant   
99                     Beer Bar              Restaurant   
100               Burrito Place              Skate Park   
101                 Men's Store           Metro Station   
102               Tanning Salon            Burger Joint   

         8th Most Common Venue      9th Most Common Venue  \
1                Movie Theater                      Motel   
2            French Restaurant             Farmers Market   
3    Middle Eastern Restaurant         Mexican Restaurant   
4                         Café             Sandwich Place   
7                  Music Venue                     Museum   
..                         ...                        ...   
98   Middle Eastern Restaurant         Miscellaneous Shop   
99         Martial Arts School                  Bookstore   
100         Light Rail Station                     Garden   
101         Mexican Restaurant  Middle Eastern Restaurant   
102                        Gym              Grocery Store   

    10th Most Common Venue  
1      Monument / Landmark  
2      Distribution Center  
3      

## Cluster 5

In [105]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
32      Scarborough             4.0            Playground   
83  Central Toronto             4.0            Playground   

   2nd Most Common Venue 3rd Most Common Venue     4th Most Common Venue  \
32        Massage Studio        Medical Center  Mediterranean Restaurant   
83          Tennis Court                Museum            Medical Center   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
32               Men's Store         Metro Station    Mexican Restaurant   
83  Mediterranean Restaurant           Men's Store         Metro Station   

        8th Most Common Venue      9th Most Common Venue  \
32  Middle Eastern Restaurant         Miscellaneous Shop   
83         Mexican Restaurant  Middle Eastern Restaurant   

   10th Most Common Venue  
32      Mobile Phone Shop  
83     Miscellaneous Shop